samples/swift/uidemo/ChatViewController.swift

```diff
@@ -0,0 +1,251 @@
+  }
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    
+    self.collectionView.backgroundColor = UIColor.whiteColor()
+    self.collectionView.delegate = self
+    let layout = self.collectionView.collectionViewLayout as! UICollectionViewFlowLayout
+    layout.minimumInteritemSpacing = CGFloat.max
+    layout.minimumLineSpacing = 4
+    
+    self.auth?.addAuthStateDidChangeListener { (auth, user) in
+      self.user = user
+    }
+    
+    self.sendButton.addTarget(self, action: #selector(didTapSend), forControlEvents: .TouchUpInside)
+  }
+  
+  override func viewDidAppear(animated: Bool) {
+    super.viewDidAppear(animated)
+    
+    self.auth?.signInAnonymouslyWithCompletion { (user, error) in
+      if let error = error {
+        fatalError("Sign in failed: \(error.localizedDescription)")
+      }
+      
+      self.query = self.chatReference.queryLimitedToLast(50)
+      
+      self.collectionViewDataSource = FirebaseCollectionViewDataSource(query: self.query!,
+        prototypeReuseIdentifier: ChatViewController.reuseIdentifier,
+        view: self.collectionView)
+      
+      // The initializer called above--though it takes a collection view--
+      // doesn't actually set the collection view's data source, so if
+      // we don't set it before trying to populate our view our app will crash.
+      self.collectionView.dataSource = self.collectionViewDataSource
+      
+      self.collectionViewDataSource.populateCellWithBlock { (anyCell, data) in
+        guard let cell = anyCell as? ChatCollectionViewCell else {
+          fatalError("Unexpected collection view cell class \(anyCell.self)")
+        }
+        
+        let chat = Chat(snapshot: data as! FIRDataSnapshot)!
+        
+        cell.textLabel.text = chat.text
+        
+        let leftRightPadding: CGFloat = 24
+        let rect = ChatCollectionViewCell.boundingRectForText(cell.textLabel.text ?? "",
+          maxWidth: self.view.frame.size.width)
+        let width = self.view.frame.size.width
+        let constant = max(width - rect.size.width - leftRightPadding, CGFloat.min)
+        if chat.uid == self.user?.uid ?? "" {
+          let colors = ChatCollectionViewCell.selfColors
+          cell.containerView.backgroundColor = colors.background
+          cell.textLabel.textColor = colors.text
+          cell.trailingConstraint.active = false
+          cell.leadingConstraint.constant = constant
+          cell.leadingConstraint.active = true
+        } else {
+          let colors = ChatCollectionViewCell.othersColors
+          cell.containerView.backgroundColor = colors.background
+          cell.textLabel.textColor = colors.text
+          cell.leadingConstraint.active = false
+          cell.trailingConstraint.constant = constant
+          cell.trailingConstraint.active = true
+        }
+      }
+      
+      // Unfortunately neither FirebaseArray nor FirebaseCollectionViewDataSource
```

Comment:
If its "unfortunately" and you work on it - i presume there is a tracking bug open for adding said listeners :)